pyiceberg/avro/resolver.py

```diff
@@ -233,7 +255,107 @@ def skip(self, decoder: BinaryDecoder) -> None:
         pass
 
 
-class SchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Reader]):
+class WriteSchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Writer]):
+    def schema(self, schema: Schema, expected_schema: Optional[IcebergType], result: Writer) -> Writer:
+        return result
+
+    def struct(self, struct: StructType, provided_struct: Optional[IcebergType], field_writers: List[Writer]) -> Writer:
+        if not isinstance(provided_struct, StructType):
+            raise ResolveError(f"File/write schema are not aligned for struct, got {provided_struct}")
+
+        provided_struct_positions: Dict[int, int] = {field.field_id: pos for pos, field in enumerate(provided_struct.fields)}
+
+        results: List[Tuple[Optional[int], Writer]] = []
+        iter(field_writers)
+
+        for pos, write_field in enumerate(struct.fields):
+            if write_field.field_id in provided_struct_positions:
+                results.append((provided_struct_positions[write_field.field_id], field_writers[pos]))
+            else:
+                # There is a default value
+                if isinstance(write_field, NestedField) and write_field.write_default is not None:
+                    # The field is not in the record, but there is a write default value
+                    default_writer = DefaultWriter(
+                        writer=visit(write_field.field_type, CONSTRUCT_WRITER_VISITOR), value=write_field.write_default
```

Comment:
I'm a little concerned by adding support for the write default. The problem is that we need to resolve the write default with child fields.

For example, if you have a field with type `s struct<a int, b int>` and default `Struct(0, 0)`, there are changes that can make the default invalid. For example, `ADD COLUMN s.c INT NOT NULL DEFAULT 0`. That would probably add a new `NestedField(name='c', field_type=IntType(), initial_default=0, write_default=0)` but it doesn't necessarily update the write or initial default for `s`. Those defaults need to be resolved somehow.

Reply:
I think if we want to move forward with this, we should just make sure that the write default is a primitive somewhere.

Reply:
Thanks, we can do that quite easily by saying that it only accepts an `L` that we already have in Typedef:

```python
# Represents the literal value
L = TypeVar("L", str, bool, int, float, bytes, UUID, Decimal, covariant=True)
```

I've also added a test:
```python

    with pytest.raises(pydantic_core.ValidationError) as exc_info:
        _ = NestedField(
            1,
            "field",
            StringType(),
            required=True,
            write_default=(1, "a", True)
        ),

    assert "validation errors for NestedField" in str(exc_info.value)
```

Reply:
I'm reluctant to fix this with types since I think that in the long term we will support nested data. Seems like we should just have an assertion for now.

Reply:
Ah, got it. We'll have to change it later to be more permissive, but that sounds fine to me. Thanks!